File: code/ch07-05/cards.erl
%% @author J D Eisenberg <jdavid.eisenberg@gmail.com>
%% @doc Functions for playing a card game.
%% @copyright 2013 J D Eisenberg
%% @version 0.1

-module(cards).
-export([make_deck/0, show_deck/1]).

%% @doc generate a deck of cards
make_deck() ->
  [{Value, Suit} || Value <- ["A", 2, 3, 4, 5, 6, 7, 8, 9, 10, "J", "Q", "K"],
    Suit <- ["Clubs", "Diamonds", "Hearts", "Spades"]].

show_deck(Deck) ->
  lists:foreach(fun(Item) -> io:format("~p~n", [Item]) end, Deck).
  

File: code/ch07-05/cards.ex
defmodule Cards do
  @moduledoc """ 
  Functions for simulating a deck of cards.
  
  from *Études for Elixir*, O'Reilly Media, Inc., 2013.
  Copyright 2013 by J. David Eisenberg.
  """ 
  @vsn 0.1 

  @doc """
  Create a deck of 52 tuples in the form [{"A", "Clubs"},
  {"A", "Diamonds"}...]
  """
  @spec make_deck() :: list(tuple)
  
  def make_deck() do
    lc value inlist ["A", 2, 3, 4, 5, 6, 7, 8, 9, "J", "Q", "K"],
      suit inlist ["Clubs", "Diamonds", "Hearts", "Spades"], do:
      {value, suit}
  end
end

